chips/nrf52/src/qdec.rs

```diff
@@ -0,0 +1,315 @@
+    registers: StaticRef<QdecRegisters>,
+    client: OptionalCell<&'static dyn kernel::hil::qdec::QdecClient>,
+    state: QdecState, /*correct enum declaration?*/
+    sample_rate: u32,
+}
+
+/// Qdec impl: provides the Qdec type with vital functionality including:
+/// FIRST DESIRED FUNCTIONALITY: new(arg1, arg2, ..., argN) -> define Qdec struct
+impl Qdec {
+    //TODO ok to be safe
+    const fn new() -> Qdec {
+        let qdec = Qdec {
+            registers: QDEC_BASE,
+            client: OptionalCell::empty(),
+            state: QdecState::START,
+        };
+        qdec
+    }
+
+    pub fn set_pins(&self, pin_a: pinmux::Pinmux, pin_b: pinmux::Pinmux) {
+         let regs = self.registers;
+        regs.psel_a.write(
+            PinSelect::Pin.val(pin_a.into()) + PinSelect::Port.val(0) + PinSelect::Connect.val(0),
+        );
+        regs.psel_b.write(
+            PinSelect::Pin.val(pin_b.into()) + PinSelect::Port.val(0) + PinSelect::Connect.val(0),
+        );
+    }
+
+    pub fn set_client(&self, client: &'static dyn kernel::hil::qdec::QdecClient) {
+        self.client.set(client);
+    }
+
+    /// When an interrupt occurs, check to see if any
+    /// of the interrupt register bits are set. If it
+    /// is, then put it in the client's bitmask
+    /// TODO: DO I NEED TO DISABLE INTERRUPTS
+    /// TODO: ADD CLIENT CODE TO HIL/CAPSULE/LIB
+    pub fn handle_interrupt(&self) {
+        self.disable_interrupts();
+        let regs = &*self.registers;
+        self.client.map(|client| {
+            let mut val = 0;
+            // For each of 4 possible compare events, if it's happened,
+            // clear it and sort its bit in val to pass in callback
+            for i in 0..4 { // TODO: either add events_compare or add each individual register
+                if self.registers.events_arr[i].is_set(Event::READY) {
+                    val = val | 1 << i;
+                    self.registers.events_arr[i].write(Event::READY::CLEAR);
+                    // Disable corresponding interrupt
+                    let interrupt_bit = match i {
+                        0 => Inte::SAMPLERDY::SET,
+                        //1 => Inte::REPORTRDY::SET,
+                        //2 => Inte::ACCOF::SET,
+                        //3 => Inte::DBLRDY::SET,
+                        4 => Inte::STOPPED::SET,
+                        _ => panic!("Unknown interrupt value!"),
+                    };
+                    if i == 0 {
+                        client.sample_ready (); 
+                    } else if i == 4 && self.qdec.state == SAMPLE_RATE_STOP {
+                        debug!("Received stopped signal!");
+                        regs.sample_per.write(SampPer::SAMPLEPER.val(5));
+                        regs.task_start.write(Task::ENABLE::SET);
+                    }
+                }
+            }
+        });
+        self.enable_interrupts();
+    }
+
+    // NOTE: ALL INTERRUPTS ARE ONLY FOR SAMPLING RIGHT NOW 
```

Comment:
Isn't this comment outdated because interrupts are also supported for changing the sample rate?